src/python/pants/java/nailgun_client.py

```diff
@@ -225,6 +229,33 @@ def try_connect(self):
     else:
       return sock
 
+  class RemoteClientStatus(enum(['before-started', 'started', 'completed'])): pass
+
+  def remote_client_status(self):
+    remote_pid = self._maybe_last_pid()
+    if remote_pid is None:
+      return self.RemoteClientStatus('before-started')
+    else:
+      if psutil.pid_exists(remote_pid):
+        return self.RemoteClientStatus.started
+      else:
+        return self.RemoteClientStatus.completed
+
+  def shutdown_gracefully(self, interval_sec=1):
+    # TODO: raise if the client hasn't started yet!
+    if self.remote_client_status() != self.RemoteClientStatus.started:
+      return
+    self.maybe_send_signal(signal.SIGINT)
+    time.sleep(interval_sec)
```

Comment:
This unconditionally sleeps quite a bit - can we somehow `select` or poll on the liveness of the thing?

Reply:
If we are claiming to solve #7014 in this PR, we should almost definitely do something like that. This is being done from a signal handler and I don't know what constraints that places on what we can do here.

Reply:
I'm just scared that if a nailgun process is blocking, this will add 3 seconds of shutdown time to pants... Probably not the end of the world for now, so I guess let's leave this for another day

Reply:
3 seconds of shutdown in the cases where it currently doesn't shut down at all though!

Reply:
Aha - I thought this was currently a case where we'd ignore and shutdown leaving a hanging nailgun process, rather than cases we wouldn't shut down :) If it's as you say, I'm cool with it!

Reply:
Oh! No, you were right. I will address the rest of your comments and then evaluate if we can do the `select` or poll to avoid a regression here.